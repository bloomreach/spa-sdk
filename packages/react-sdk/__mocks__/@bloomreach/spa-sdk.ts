/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const spaSdk = jest.requireActual('@bloomreach/spa-sdk');

export const META_POSITION_BEGIN = spaSdk.META_POSITION_BEGIN;
export const META_POSITION_END = spaSdk.META_POSITION_END;

export function mockMeta(data: string, position: string) {
  return {
    getData: jest.fn().mockReturnValue(data),
    getPosition: jest.fn().mockReturnValue(position),
  };
}

export function mockNoCommentMeta() {
  return mockMeta('not-a-comment', META_POSITION_BEGIN);
}

const componentMock = {
  getMeta: jest.fn().mockReturnValue([
    mockMeta('meta-begin', META_POSITION_BEGIN),
    mockMeta('meta-end', META_POSITION_END),
  ]),
};

export const pageMock = {
  getComponent: jest.fn().mockReturnValue(componentMock),
  getContent: jest.fn(),
  getTitle: jest.fn(),
  sync: jest.fn(),
};

export const initialize = jest.fn();
export const destroy = jest.fn();

export function isMetaComment(value: any) {
  return value.getData() !== 'not-a-comment';
}
