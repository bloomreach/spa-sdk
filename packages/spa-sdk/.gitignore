/coverage
/docs
/lib
/node_modules
.DS_store