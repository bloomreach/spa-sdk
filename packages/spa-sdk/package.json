{
  "name": "@bloomreach/spa-sdk",
  "version": "14.0.0",
  "description": "Bloomreach SPA SDK",
  "keywords": [
    "bloomreach",
    "spa",
    "sdk"
  ],
  "author": "Bloomreach B.V.",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://code.onehippo.org/cms-community/bloomreach-spa-sdk",
    "directory": "packages/spa-sdk"
  },
  "bugs": {
    "url": "https://issues.onehippo.com/projects/CMS/issues"
  },
  "main": "dist/index.js",
  "browser": "dist/index.js",
  "module": "dist/index.mjs",
  "es2015": "dist/index.es6.mjs",
  "types": "dist/index.d.ts",
  "files": [
    "dist/"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js",
    "docs": "tsc --emitDeclarationOnly && dts-bundle-generator --out-file dist/spa-sdk.d.ts dist/index.d.ts && typedoc dist/spa-sdk.d.ts && minicat src/docs/typedoc.css >> docs/assets/css/main.css && rm dist/spa-sdk.d.ts",
    "lint": "tslint --project .",
    "test": "jest"
  },
  "jest": {
    "testEnvironment": "jsdom",
    "coverageDirectory": "./coverage",
    "collectCoverage": true,
    "collectCoverageFrom": [
      "src/**/*.ts"
    ],
    "moduleFileExtensions": [
      "ts",
      "js"
    ],
    "transform": {
      "^.+\\.ts$": "ts-jest"
    },
    "globals": {
      "ts-jest": {
        "tsConfig": "tsconfig.json"
      }
    },
    "testMatch": [
      "**/?(*.)spec.ts"
    ]
  },
  "devDependencies": {
    "@babel/core": "^7.6",
    "@babel/plugin-proposal-class-properties": "^7.5",
    "@babel/plugin-proposal-object-rest-spread": "^7.5",
    "@babel/plugin-transform-object-assign": "^7.2",
    "@babel/polyfill": "^7.6",
    "@babel/preset-env": "^7.6",
    "@types/jest": "^24.0",
    "@types/node": "^12.0",
    "babel-loader": "^8.0",
    "babel-plugin-transform-async-to-promises": "^0.8",
    "dts-bundle-generator": "^3.2",
    "jest": "^24.0",
    "jest-environment-jsdom-fifteen": "^1.0",
    "js-beautify": "^1.0",
    "minicat": "^1.0",
    "rollup": "^1.21",
    "rollup-plugin-babel": "^4.3",
    "rollup-plugin-terser": "^5.1",
    "rollup-plugin-typescript2": "^0.24",
    "ts-jest": "^24.0",
    "ts-loader": "^5.0",
    "tslint": "^5.0",
    "tslint-config-airbnb": "^5.0",
    "tslint-loader": "^3.5",
    "typedoc": "^0.15.0",
    "typescript": "^3.6"
  },
  "dependencies": {
    "emittery": "^0.4"
  }
}
