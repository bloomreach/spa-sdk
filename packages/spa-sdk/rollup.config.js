/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import babel from 'rollup-plugin-babel';
import dts from 'rollup-plugin-dts';
import { terser } from 'rollup-plugin-terser';
import typescript from 'rollup-plugin-typescript2';

export default [
  {
    input: 'src/index.ts',
    output: [
      {
        exports: 'named',
        file: 'dist/spa-sdk.js',
        format: 'umd',
        name: 'BloomreachSpaSdk',
      },
      {
        file: 'dist/spa-sdk.mjs',
        format: 'esm',
      },
    ],
    plugins: [
      typescript({ cacheRoot: './node_modules/.cache/rpt2' }),
      babel({ extensions: ['.ts'] }),
      terser({
        ecma: 5,
        mangle: false,
        compress: false,
        output: {
          beautify: true,
          comments: false,
        },
        sourcemap: false,
      })
    ],
  },

  {
    input: 'src/index.ts',
    output: [{
      file: 'dist/spa-sdk.es6.mjs',
      format: 'esm',
    }],
    plugins: [
      typescript({ cacheRoot: './node_modules/.cache/rpt2' }),
      terser({
        ecma: 5,
        mangle: false,
        compress: false,
        output: {
          beautify: true,
          comments: false,
        },
        sourcemap: false,
      })
    ],
  },

  {
    input: 'src/index.ts',
    output: [
      {
        file: 'dist/spa-sdk.d.ts',
        format: 'es',
      },
    ],
    plugins: [ dts() ],
  }
];
