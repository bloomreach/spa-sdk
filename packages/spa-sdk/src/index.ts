/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Main entry point of the spa-sdk library.
 * @module index
 */

import { Typed } from 'emittery';
import { Events } from './events';
import { Cms } from './cms';
import {
  ComponentFactory,
  ComponentImpl,
  ContainerItemImpl,
  ContainerItemModel,
  ContainerItem,
  ContainerImpl,
  ContainerModel,
  ContentImpl,
  LinkFactory,
  MetaCommentImpl,
  MetaFactory,
  PageImpl,
  Page,
  SingleTypeFactory,
  TYPE_COMPONENT,
  TYPE_COMPONENT_CONTAINER_ITEM,
  TYPE_COMPONENT_CONTAINER,
  TYPE_META_COMMENT,
  TYPE_LINK_EXTERNAL,
  TYPE_LINK_INTERNAL,
  TYPE_LINK_RESOURCE,
} from './page';
import { Configuration, Spa } from './spa';
import { UrlBuilderImpl } from './url';

const eventBus = new Typed<Events>();
const cms = new Cms(eventBus);
const pages = new WeakMap<Page, Spa>();

/**
 * Initializes the page model.
 *
 * @param config Configuration of the SPA integration with brXM.
 */
export async function initialize(config: Configuration): Promise<Page> {
  const urlBuilder =  new UrlBuilderImpl();
  const linkFactory = new LinkFactory()
    .register(TYPE_LINK_EXTERNAL, urlBuilder.getSpaUrl.bind(urlBuilder))
    .register(TYPE_LINK_INTERNAL, urlBuilder.getSpaUrl.bind(urlBuilder))
    .register(TYPE_LINK_RESOURCE, urlBuilder.getCmsUrl.bind(urlBuilder));
  const metaFactory = new MetaFactory()
    .register(TYPE_META_COMMENT, (model, position) => new MetaCommentImpl(model, position));
  const componentFactory = new ComponentFactory()
    .register(TYPE_COMPONENT, (model, children) => new ComponentImpl(model, children, linkFactory, metaFactory))
    .register(TYPE_COMPONENT_CONTAINER, (model, children) => new ContainerImpl(
      model as ContainerModel,
      children as ContainerItem[],
      linkFactory,
      metaFactory,
    ))
    .register(TYPE_COMPONENT_CONTAINER_ITEM, model => new ContainerItemImpl(
      model as ContainerItemModel,
      eventBus,
      linkFactory,
      metaFactory,
    ));
  const contentFactory = new SingleTypeFactory(model => new ContentImpl(model, linkFactory, metaFactory));
  const pageFactory = new SingleTypeFactory(model => new PageImpl(
    model,
    componentFactory.create(model.page),
    contentFactory,
    eventBus,
    linkFactory,
    metaFactory,
  ));

  const spa = new Spa(config, cms, eventBus, pageFactory, urlBuilder);
  const page = await spa.initialize();

  pages.set(page, spa);

  return page;
}

/**
 * Destroys the integration with the SPA page.
 * @param page Page instance to destroy.
 */
export function destroy(page: Page) {
  const spa = pages.get(page);

  return spa && spa.destroy();
}

export { Configuration } from './spa';
export {
  Component,
  ContainerItem,
  Container,
  Content,
  Link,
  Menu,
  MetaComment,
  Meta,
  Page,
  isComponent,
  isContainerItem,
  isContainer,
  isLink,
  isMetaComment,
  isMeta,
  META_POSITION_BEGIN,
  META_POSITION_END,
  TYPE_CONTAINER_BOX,
  TYPE_CONTAINER_INLINE,
  TYPE_CONTAINER_NO_MARKUP,
  TYPE_CONTAINER_ORDERED_LIST,
  TYPE_CONTAINER_UNORDERED_LIST,
  TYPE_LINK_EXTERNAL,
  TYPE_LINK_INTERNAL,
  TYPE_LINK_RESOURCE,
} from './page';
