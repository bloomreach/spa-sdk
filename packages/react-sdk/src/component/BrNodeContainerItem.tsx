/*
 * Copyright 2019-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useCallback, useContext, useEffect, useState } from 'react';
import { ContainerItem, TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { BrContainerItemUndefined } from '../cms';
import { BrProps } from './BrProps';
import { BrMappingContext } from './BrMappingContext';
import { BrMeta } from '../meta';

export function BrNodeContainerItem(
  props: React.PropsWithChildren<BrProps<ContainerItem>>,
): React.ReactElement {
  const context = useContext(BrMappingContext);
  const { component, page, children } = props;
  const [forcedRerenders, setForcedRerenders] = useState(0);

  const onUpdate = useCallback((): void => {
    // Trigger a rerender of this component due to a non-react 3rd party event
    setForcedRerenders((c) => c + 1);
  }, []);

  useEffect(() => {
    component?.on('update', onUpdate);

    return () => {
      component?.off('update', onUpdate);
    };
  }, [component, onUpdate]);

  useEffect(() => {
    // Skip the mount event page sync as BrPage just called it
    if (forcedRerenders > 0) {
      page?.sync();
    }
  }, [page, forcedRerenders]);

  const getMapping = (): React.ComponentType<BrProps> => {
    const type = component?.getType();

    if (type && type in context) {
      return context[type] as React.ComponentType<BrProps>;
    }

    return (
      (context[
        TYPE_CONTAINER_ITEM_UNDEFINED as any
      ] as React.ComponentType<BrProps>) ?? BrContainerItemUndefined
    );
  };

  const mapping = getMapping();
  const meta = component?.getMeta();
  const content = mapping ? React.createElement(mapping, props) : children;

  return React.createElement(BrMeta, { meta }, content);
}
