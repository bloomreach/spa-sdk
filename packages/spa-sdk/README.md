# Bloomreach SPA SDK
[![NPM](https://img.shields.io/npm/v/@bloomreach/spa-sdk.svg)](https://www.npmjs.com/package/@bloomreach/spa-sdk)
[![License](https://img.shields.io/npm/l/@bloomreach/spa-sdk.svg)](http://www.apache.org/licenses/LICENSE-2.0)

Bloomreach SPA SDK provides simplified headless integration with [Bloomreach Experience Manager](https://www.bloomreach.com/en/products/experience-manager) for JavaScript-based applications.
This library interacts with the [Page Model API](https://documentation.bloomreach.com/library/concepts/page-model-api/introduction.html) and exposes a simplified and framework-agnostic interface over the page model.

## What is Bloomreach Experience Manager?
Bloomreach Experience Manager (brXM) is an open and flexible CMS designed for developers and marketers. As the original headless CMS, brXM allows developers to build quickly and integrate with the systems. While itâ€™s built for speed, it also provides top-notch personalization and channel management capabilities for marketers to drive results.

## Features
- [Page Model API](https://documentation.bloomreach.com/library/concepts/page-model-api/introduction.html) client;
- Page Model parser;
- URL generator;
- brXM integration.

## Get Started
### Installation
To get the SDK into your project with [NPM](https://docs.npmjs.com/cli/npm):
```bash
npm install @bloomreach/spa-sdk
```

And with [Yarn](https://yarnpkg.com):
```bash
yarn add @bloomreach/spa-sdk
```

### Usage
The following code snippet requests a related page model and shows the page's title.

```javascript
import axios from 'axios';
import { initialize } from '@bloomreach/spa-sdk';

async function showPage(path) {
  const page = await initialize({
    httpClient: axios,
    options: {
      live: {
        cmsBaseUrl: 'http://localhost:8080/site',
        spaBaseUrl: '',
      },
      preview: {
        cmsBaseUrl: 'http://localhost:8080/site/_cmsinternal',
        spaBaseUrl: '/site/_cmsinternal?bloomreach-preview=true',
      },
    },
    request: { path },
  });

  document.querySelector('#title').innerText = page.getTitle();
}

showPage(`${window.location.pathname}${window.location.search}`);
```

### Configuration
The `initialize` function supports several options you may use to customize page initialization.

Option | Required | Default | Description
--- | :---: | --- | ---
`httpClient` | yes | _none_ | The HTTP client that will be used to fetch the page model. Signature is similar to [Axios](https://github.com/axios/axios#axiosconfig) client.
`options` | yes | _none_ | The CMS URL options.
`options.live` | yes | _none_ | The CMS URL options for the live site.
`options.live.apiBaseUrl` | no | `options.live.cmsBaseUrl` + `"/resourceapi"` | Base URL of the Page Model API for the live site (e.g. `http://localhost:8080/site/resourceapi` or `http://localhost:8080/site/channel/resourceapi`).
`options.live.cmsBaseUrl` | yes | _none_ | Base URL of the live site (e.g. `http://localhost:8080/site` or `http://localhost:8080/site/channel`).
`options.live.spaBaseUrl` | no | `""` | Base URL of the live SPA (e.g. `/account` or `//www.example.com`).
`options.preview` | yes | _none_ | The CMS URL options for the preview site.
`options.preview.apiBaseUrl` | no | `options.preview.cmsBaseUrl` + `"/resourceapi"` | Base URL of the Page Model API for the preview site (e.g. `http://localhost:8080/site/_cmsinternal/resourceapi` or `http://localhost:8080/site/_cmsinternal/channel/resourceapi`).
`options.preview.cmsBaseUrl` | yes | _none_ | Base URL of the live site (e.g. `http://localhost:8080/site/_cmsinternal` or `http://localhost:8080/site/_cmsinternal/channel`).
`options.preview.spaBaseUrl` | no | `""` | Base URL of the live SPA (e.g. `/site/_cmsinternal?bloomreach-preview=true` or `/site/_cmsinternal/channel?bloomreach-preview=true`). This path and query string parameters will be used to detect whether it is a preview mode or not.
`request` | yes | _none_ | Current user's request.
`request.path` | yes | _none_ | The path part of the URL, including a query string if present (e.g. `/path/to/page?foo=1`).
`request.headers` | no | `{}` | An object holding request headers. It should contain a `Cookie` header if rendering is happening on the server-side.
`window` | no | `window` | A window object reference will be used to interact with brXM. It needs to be set when initialize is being called within an iframe or worker process.

### Reference
The complete API reference can be found [here](https://javadoc.onehippo.org/14.0/bloomreach-spa-sdk/).

#### Functions
Function | Description
--- | ---
`initialize(config, model?): Promise<Page>` | This function accepts a configuration object as an argument and returns a promisified JavaScript object representing the page model. In case, when the page model has already been fetched, you can pass this JSON blob as a second parameter.
`destroy(page: Page): void` | Destroys the integration with the SPA page object.
`isPage(value): boolean` | Checks whether a value is a page.
`isComponent(value): boolean` | Checks whether a value is a page component.
`isContainer(value): boolean` | Checks whether a value is a page container.
`isContainerItem(value): boolean` | Checks whether a value is a page container item.
`isMeta(value): boolean` | Checks whether a value is a meta-data object.
`isMetaComment(value): boolean` | Checks whether a value is a meta-data comment.
`isLink(value): boolean` | Checks whether a value is a link.
`isReference(value): boolean` | Checks whether a value is a content reference.

#### Constants
Constant | Description
--- | ---
`META_POSITION_BEGIN` | Meta-data following before a page component.
`META_POSITION_END` | Meta-data following after a page component.
`TYPE_CONTAINER_BOX` | A blocked container with blocked items.
`TYPE_CONTAINER_INLINE` | A blocked container with inline items.
`TYPE_CONTAINER_NO_MARKUP` | A container without surrounding markup.
`TYPE_CONTAINER_ORDERED_LIST` | An ordered list container.
`TYPE_CONTAINER_UNORDERED_LIST` | An unordered list container.
`TYPE_LINK_EXTERNAL` | Link to a page outside the current application.
`TYPE_LINK_INTERNAL` | Link to a page inside the current application.
`TYPE_LINK_RESOURCE` | Link to a CMS resource.

#### Objects
##### Page
The `Page` class represents the brXM page to render. This is the main entry point to the page model.

Method | Description
--- | ---
<code>getComponent(...componentNames): Component &vert; undefined</code> | Gets a component in the page (e.g. `getComponent('main', 'right')`). If `componentNames` is omitted, then the page root component will be returned.
<code>getContent(reference: Reference &vert; string): Content &vert; undefined</code> | Gets a content item used on the page.
`getMeta(meta): Meta[]` | Generates a meta-data from the provided `meta` model.
<code>getTitle(): string &vert; undefined</code> | Gets the title of the page, or `undefined` if not configured.
`getUrl(link?: Link): string` | Generates a URL for a link object.<br> - If the link object type is internal, then it will prepend `spaBaseUrl`. In case when the link starts with the same path as in `cmsBaseUrl`, this part will be removed.<br> - If the link parameter is omitted, then the link to the current page will be returned.<br> - In other cases, the link will be returned as-is.
`getUrl(path: string): string` | Generates an SPA URL for the path.<br> - If it is a relative path, then it will prepend `spaBaseUrl`.<br> - If it is an absolute path, then the behavior will be similar to internal link generation.
`isPreview(): boolean` | Returns whether the page is in the preview mode.
`rewriteLinks(content: string, type?: string): string` | Rewrite links to pages and resources in the HTML content. This method looks up for `a` tags with `data-type` and `href` attributes and `img` tags with `src` attribute. Links will be updated according to the configuration used to initialize the page. The `type` parameter is similar to `mimeType` parameter of the [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser).
`sync(): void` | Synchronizes the brXM integration state (UI elements positions).
`toJSON(): object` | A plain JavaScript object of the page model.

##### Component
The `Component` class corresponds to page nodes, and it may hold other components inside.

Method | Description
--- | ---
`getId(): string` | Returns the component id.
`getMeta(): Meta[]` | Returns the component meta-data collection.
`getModels(): object` | Returns the map of the component models.
<code>getUrl(): string &vert; undefined</code> | Returns the link to the partial component model.
`getName(): string` | Returns the name of the component.
`getParameters(): object` | Returns the parameters of the component.
`getChildren(): Component[]` | Returns the direct children of the component.
<code>getComponent(...componentNames: string[]): Component &vert; undefined</code> | Looks up for a nested component.
<code>getComponentById(id: string): Component &vert; undefined</code> | Looks up for a nested component by its id.

##### Container
The `Container` class represents a page node that is actually present in the DOM but as an element surrounding its children. Container extends the Component class, and therefore, all the [Component methods](#component) are applicable here.

Method | Description
--- | ---
<code>getType(): string &vert; undefined</code> | Returns the [type](https://documentation.bloomreach.com/library/concepts/template-composer/channel-editor-containers.html) of a container.

##### Container Item
The `ContainerItem` objects are usually visible on the page and interact with the user. Container Item extends the Component class, and therefore, all the [Component methods](#component) are applicable here.

Method | Description
--- | ---
<code>getType(): string &vert; undefined</code> | Returns the type of a container item. The available types depend on which container items have been configured in the backend (e.g. "Banner").
`isHidden(): boolean` | Returns whether the component should not render anything. Hiding components is only possible with the Relevance feature.
`on(eventName: string, listener: Function): Function` | Subscribes for an event and returns the unsubscribe function.
`off(eventName: string, listener: Function): void` | Unsubscribes from an event.

##### Content
The `Content` object holds document data that is used by the page components.

Method | Description
--- | ---
`getId(): string` | Returns the content id.
<code>getLocale(): string &vert; undefined</code> | Returns the content locale.
`getMeta(): Meta[]` | Returns the content meta-data collection.
`getName(): string` | Returns the content name.
`getData(): object` | Returns the content data as it is returned in the Page Model API.
<code>getUrl(): string &vert; undefined</code> | Returns the link to the content.

##### Meta
The `Meta` objects are being used by the brXM to page and its components.

Method | Description
--- | ---
`getData(): string` | Returns the meta-data.
`getPosition(): string` | Returns the meta-data position relative to the related element.

## Links
- [SPA integration concept](https://documentation.bloomreach.com/library/concepts/spa-integration/introduction.html).
- [Page Model API introduction](https://documentation.bloomreach.com/library/concepts/page-model-api/introduction.html).

## FAQ
- Information about common problems and possible solutions can be found on [the troubleshooting page](https://documentation.bloomreach.com/library/concepts/spa-integration/troubleshooting.html).
- Information about the recommended setup can be found on [the best practices page](https://documentation.bloomreach.com/library/concepts/spa-integration/best-practices.html).

## License
Published under [Apache 2.0](http://www.apache.org/licenses/LICENSE-2.0) license.
