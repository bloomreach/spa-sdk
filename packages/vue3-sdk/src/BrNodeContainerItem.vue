<!--
  Copyright 2023 Bloomreach

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

    https://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
  -->

<template>
  <Suspense>
    <component
      :is="mapping[componentType]"
      v-if="componentType && componentType in mapping"
      :component="component"
      :page="page"
    />

    <component
      :is="mapping[TYPE_CONTAINER_ITEM_UNDEFINED]"
      v-else-if="TYPE_CONTAINER_ITEM_UNDEFINED in mapping"
      :component="component"
      :page="page"
    />

    <br-container-item-undefined v-else-if="component" :component="component"/>
  </Suspense>
</template>

<script setup lang="ts">
import { component$, mapping$, page$ } from '@/providerKeys';
import type { ContainerItem } from '@bloomreach/spa-sdk';
import { TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { computed, inject } from 'vue';
import BrContainerItemUndefined from './BrContainerItemUndefined.vue';

const page = inject(page$);
const mapping = inject(mapping$, {});
const component = inject<ContainerItem>(component$);
const componentType = computed(() => component?.getType());
</script>
