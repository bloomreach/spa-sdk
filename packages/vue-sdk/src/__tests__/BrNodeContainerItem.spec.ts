/*
 * Copyright 2020-2025 Bloomreach
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Component, ContainerItem, Page } from '@bloomreach/spa-sdk';
import { TYPE_CONTAINER_ITEM_UNDEFINED } from '@bloomreach/spa-sdk';
import { shallowMount } from '@vue/test-utils';
import type { GlobalMountOptions } from '@vue/test-utils/dist/types';
import type { Mocked } from 'vitest';
import { beforeAll, describe, expect, it, vi } from 'vitest';
import type { PropType, Ref } from 'vue';
import { defineComponent, nextTick, ref } from 'vue';
import type { BrMapping } from '../../typings';
import BrNodeContainerItem from '@/BrNodeContainerItem.vue';
import { component$, mapping$, page$ } from '../providerKeys';

describe('BrNodeContainerItem', () => {
  let component: Mocked<ContainerItem>;
  let componentRef: Ref<typeof component>;
  let mapping: BrMapping;
  let page: Mocked<Page>;

  let global: GlobalMountOptions;

  const SomeContainerItem = defineComponent({
    name: 'some-container-item',
    template: `
      <div class="some-container-item"/>
    `,
    props: {
      page: {
        type: Object as PropType<Page>,
      },
      component: {
        type: Object as PropType<Component>,
      },
    },
  });

  beforeAll(() => {
    component = {
      getType: vi.fn(),
      on: vi.fn(),
    } as unknown as typeof component;
    componentRef = ref(component);
    mapping = { something: SomeContainerItem };
    page = { sync: vi.fn() } as unknown as typeof page;

    global = {
      provide: {
        [page$ as symbol]: ref(page),
        [mapping$ as symbol]: mapping,
        [component$ as symbol]: componentRef,
      },
    };
  });

  describe('render', () => {
    it('should render a mapped container item', async () => {
      component.getType.mockReturnValue('something');

      const wrapper = shallowMount(BrNodeContainerItem, { global });
      await nextTick();

      const props = wrapper.findComponent(SomeContainerItem).props();

      expect(wrapper.html()).toMatchSnapshot();
      expect(props.component).toEqual(component);
      expect(props.page).toEqual(page);
    });

    it('should render an undefined container item', async () => {
      component.getType.mockReturnValue('undefined');

      const wrapper = shallowMount(BrNodeContainerItem, { global });
      await nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should override undefined container item', async () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      mapping[TYPE_CONTAINER_ITEM_UNDEFINED as any] = SomeContainerItem;
      component.getType.mockReturnValue('undefined');

      const wrapper = shallowMount(BrNodeContainerItem, { global });
      await nextTick();

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('destroyed', () => {
    it('should unsubscribe from update event on component destroy', async () => {
      const unsubscribe = vi.fn();
      component.on.mockReturnValue(unsubscribe);

      const wrapper = shallowMount(BrNodeContainerItem, { global });
      await nextTick();
      wrapper.unmount();
      await nextTick();

      expect(unsubscribe).toBeCalled();
    });
  });

  describe('update', () => {
    it('should subscribe for update event', async () => {
      shallowMount(BrNodeContainerItem, { global });
      await nextTick();

      expect(component.on).toBeCalledWith('update', expect.any(Function));
    });

    it('should unsubscribe from update event on component change', async () => {
      const unsubscribe = vi.fn();
      component.on.mockReturnValue(unsubscribe);

      shallowMount(BrNodeContainerItem, { global });
      await nextTick();

      componentRef.value = { ...component };
      await nextTick();

      expect(unsubscribe).toBeCalled();
    });

    it('should sync page on update event', async () => {
      shallowMount(BrNodeContainerItem, { global });
      await nextTick();

      const [[, listener]] = component.on.mock.calls;
      listener({});

      expect(page.sync).toBeCalled();
    });
  });
});
