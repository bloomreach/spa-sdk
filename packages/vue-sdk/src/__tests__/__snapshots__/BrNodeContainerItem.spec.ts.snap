// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BrNodeContainerItem > render > should override undefined container item 1`] = `
"<br-meta-stub>
  <some-container-item-stub component=\\"[object Object]\\" page=\\"[object Object]\\"></some-container-item-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainerItem > render > should render a mapped container item 1`] = `
"<br-meta-stub>
  <some-container-item-stub component=\\"[object Object]\\" page=\\"[object Object]\\"></some-container-item-stub>
</br-meta-stub>"
`;

exports[`BrNodeContainerItem > render > should render an undefined container item 1`] = `
"<br-meta-stub>
  <br-container-item-undefined-stub component=\\"[object Object]\\"></br-container-item-undefined-stub>
</br-meta-stub>"
`;
